### PR TITLE
Remove synchronization in Settings#keySet() and Settings#names()

### PR DESCRIPTION
**Pull request by probakowski**
Currently every invocation of Settings#keySet() and Settings#names() is synchronized even though we access immutable set. 
This change makes use of AtomicReferences instead and removes need for synchronization and avoids its cost.

### CODE REVIEW

**Review by rjernst — COMMENTED** [review]
I wonder if we could keep the SetOnce but add a set check around it? The synchronized block can still be inside the null condition, so that we don't duplicate the work to create these keysets. It's not that I think the sets are too costly to create, just that I like the SetOnce protection.

**Comment by probakowski** [comment]
@rjernst we could add check as you suggested but that still leave as with synchronization that is slower than creating sets (~2x for uncongested access and more than 2 orders of magnitude with heavy congestion). 
If we use correct concurrent implementation of `SetOnce` we can skip synchronized block (code is simpler, too) and still have guarantee that only single set is ever returned. Unfortunately there is race condition in implementation from Lucene (https://issues.apache.org/jira/browse/LUCENE-9001), so I've added new implementation to our codebase.
What do you think?

**Comment by rjernst** [comment]
I'm ok with that general approach. Thanks for opening an issue in Lucene. We should have appropriate documentation on the fork of SetOnce pointing to that issue, as well as a deprecated annotation (every use of it we add must be converted once the Lucene issue is resolved, so we should be careful about where it is really needed). I would also like @jasontedor to have a look here.

**Comment by probakowski** [comment]
Thanks @rjernst, I've added deprecation message pointing to [LUCENE-9001](https://issues.apache.org/jira/browse/LUCENE-9001).
@jasontedor your input would be greatly appreciated.

File: server/src/main/java/org/elasticsearch/common/settings/Settings.java [hunk 1]
```diff
@@ -506,9 +493,7 @@ public Version getAsVersion(String setting, Version defaultVersion) throws Setti
      */
     public String toDelimitedString(char delimiter) {
         StringBuilder sb = new StringBuilder();
-        for (Map.Entry<String, Object> entry : settings.entrySet()) {
-            sb.append(entry.getKey()).append("=").append(entry.getValue()).append(delimiter);
-        }
+        settings.forEach((key, value) -> sb.append(key).append("=").append(value).append(delimiter));
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/common/settings/Settings.java]: This appears to be an unrelated formatting change. I’m also doubting it, since now we have those lambda allocations in a loop. I’ve personally never been a fan of using `forEach` to mutate, I’ve felt it should be side-effect free (because it looks like an expression, not a statement). Either way, it’s unrelated to our main story.

**Review by jasontedor — COMMENTED** [review]
I’m okay with the change, but I do want to say I would never expect a Settings object to be under heavy contention, so it’s mostly a distinction without a difference. If we feel strongly about making this change, I think that we should wait for Lucene upstream to incorporate the SetOnce fix (if they do, I see there’s a little discussion about it), rather than committing our own here.

**Comment by rjernst** [comment]
@probakowski I think the main contention concern here should be addressed by https://github.com/elastic/elasticsearch/pull/52467? Can this be closed?

**Comment by jasontedor** [comment]
@rjernst I think that #52435 already took care of this.
